test/workers/delete_object_hierarchy_worker_test.rb

```diff
@@ -141,23 +223,237 @@ class DeleteServiceTest < DeleteObjectHierarchyWorkerTest
-      DeleteObjectHierarchyWorker.stubs(:perform_later)
+    def all_objects
+      base_models.inject(Set.new) do |acc, model|
+        acc.merge model.all
+      end
+    end
+  end
+
+  class DeleteAccountTest < ActiveSupport::TestCase
+    include ActiveJob::TestHelper
+
+    attr_reader :provider
+
+    setup do
+      @provider = FactoryBot.create(:provider_account, :with_a_buyer)
+    end
+
+    test "delete succeeds" do
+      provider2 = FactoryBot.create(:provider_account, :with_a_buyer)
+
+      buyer1 = provider.buyers.first
+      buyer2 = FactoryBot.create(:buyer_account, provider_account: provider)
+      provider.schedule_for_deletion!
+      perform_enqueued_jobs(queue: "deletion") do
+        DeleteObjectHierarchyWorker.delete_later buyer1
+      end
+      assert_raise(ActiveRecord::RecordNotFound) { buyer1.reload }
+      buyer2.reload
+
+      perform_enqueued_jobs(queue: "deletion") do
+        DeleteObjectHierarchyWorker.delete_later provider
+      end
+
+      assert_raise(ActiveRecord::RecordNotFound) { provider.reload }
+      assert_raise(ActiveRecord::RecordNotFound) { buyer2.reload }
+
+      provider2.reload
+      assert_equal 1, provider2.buyers.count
+    end
+
+    test "delete with a payment_gateway_setting" do
+      pgs = FactoryBot.create(:payment_gateway_setting, account: provider)
+      provider.schedule_for_deletion!
+      perform_enqueued_jobs(queue: "deletion") do
+        DeleteObjectHierarchyWorker.delete_later provider
+      end
+
+      assert_raise(ActiveRecord::RecordNotFound) { provider.reload }
+      assert_raise(ActiveRecord::RecordNotFound) { pgs.reload }
+    end
+
+    test "account is not deleted when not scheduled" do
+      buyer = provider.buyers.first
+      perform_enqueued_jobs(queue: "deletion") do
+        assert_raises { DeleteObjectHierarchyWorker.delete_later buyer }
+        assert_raises { DeleteObjectHierarchyWorker.delete_later provider }
+      end
+      assert provider.reload
+      assert buyer.reload
+    end
+
+    class DeleteCompleteAccountTest < ActiveSupport::TestCase
+      include ActiveJob::TestHelper
+      include TestHelpers::Provider
+
+      # DeletedObject might be needed for updating backend, JanitorWorker handles stale ones
+      # the rest are not specific to a provider or should not be deleted with the provider
+      NON_PROVIDER_MODELS = [BackendEvent, LogEntry, Country, DeletedObject]
+
+      test 'perform big account destroy in background' do
+        provider = create_a_complete_provider
+        assert_equal 1, Account.where(provider: true).count
+        assert_equal 1, Account.where(buyer: true).count
+
+        provider.schedule_for_deletion!
+        assert_empty models_without_objects.map(&:to_s)
+
+        perform_enqueued_jobs { DeleteObjectHierarchyWorker.delete_later(provider) }
+        assert_empty performed_jobs.find { _1["job_class"] == "DeleteObjectHierarchyWorker" }["exception_executions"]
```

Comment:
Is this assertion valid also when `performed_jobs.find { _1["job_class"] == "DeleteObjectHierarchyWorker" }` doesn't find anything? I'm not sure if `find` here returns `nil`, returns empty enumerator, raises an exception...

Reply:
returns `nil`. But if it returns `nil`, then the `[]` call will fails with `NoMethodError`. I don't find it likely though because we do perform a delete job so should be some really weird issue.

On the other hand, it is also a useless assertion, because any error should pop out into the test and be reported much earlier. I think to just remove this line.